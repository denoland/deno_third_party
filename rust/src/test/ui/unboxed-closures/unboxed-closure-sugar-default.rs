// Test interaction between unboxed closure sugar and default type
// parameters (should be exactly as if angle brackets were used).

#![feature(unboxed_closures)]
#![allow(dead_code)]

trait Foo<T,V=T> {
    type Output;
    fn dummy(&self, t: T, v: V);
}

trait Eq<X: ?Sized> { fn same_types(&self, x: &X) -> bool { true } }
impl<X: ?Sized> Eq<X> for X { }
fn eq<A: ?Sized,B: ?Sized>() where A : Eq<B> { }

fn test<'a,'b>() {
    // Parens are equivalent to omitting default in angle.
    eq::< Foo<(isize,),Output=()>,                   Foo(isize)                      >();

    // In angle version, we supply something other than the default
    eq::< Foo<(isize,),isize,Output=()>,      Foo(isize)                      >();
    //~^ ERROR E0277

    // Supply default explicitly.
    eq::< Foo<(isize,),(isize,),Output=()>,   Foo(isize)                      >();
}

fn main() { }
