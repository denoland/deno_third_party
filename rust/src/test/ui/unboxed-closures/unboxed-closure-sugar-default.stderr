error[E0277]: the trait bound `dyn Foo<(isize,), isize, Output=()>: Eq<dyn Foo<(isize,), Output=()>>` is not satisfied
  --> $DIR/unboxed-closure-sugar-default.rs:21:5
   |
LL |     eq::< Foo<(isize,),isize,Output=()>,      Foo(isize)                      >();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Eq<dyn Foo<(isize,), Output=()>>` is not implemented for `dyn Foo<(isize,), isize, Output=()>`
   |
note: required by `eq`
  --> $DIR/unboxed-closure-sugar-default.rs:14:1
   |
LL | fn eq<A: ?Sized,B: ?Sized>() where A : Eq<B> { }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
